File: lab4/lab4/Matrix.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab4
{
    public class Matrix
    {
        private int[,] matrix;
        public int Size { get { return matrix.GetLength(0); } }

        public Matrix(int size)
        {
            matrix = new int[size,size];
        }

        public void Generate(int fillDigit)
        {
            for (int i = 0; i < Size; i++)
                for (int j = 0; j < Size; j++)
                    matrix[i, j] = fillDigit;
        }

        #region Operators
        public static Matrix operator *(Matrix m1, Matrix m2)
        {
            Matrix rezult = new Matrix(m1.Size);
            int tempValue = 0;
            for (int i = 0; i < rezult.Size; i++)
                for (int j = 0; j < rezult.Size; j++)
                {
                    tempValue = 0;
                    for (int k = 0; k < rezult.Size; k++)
                        tempValue += m1.GetElement(j, k) * m2.GetElement(k, j);
                    rezult.SetElement(i, j, tempValue);
                }
            return rezult;
        }

        public static Vector operator *(Matrix matrix, Vector vector)
        {
            Vector rezult = new Vector(vector.Size);
            int tempValue = 0;
            for (int i = 0; i < rezult.Size; i++)
            {
                tempValue = 0;
                for (int j = 0; j < rezult.Size; j++)
                    tempValue += vector.GetElement(i) * matrix.GetElement(j, i);
                rezult.SetElement(i, tempValue);
            }
            return rezult;
        }

        public static Matrix operator +(Matrix m1, Matrix m2)
        {
            Matrix rezult = new Matrix(m1.Size);
            for (int i = 0; i < m1.Size; i++)
                for (int j = 0; j < m1.Size; j++)
                    rezult.SetElement(i, j, m1.GetElement(i, j) + m2.GetElement(i, j));
            return rezult;
        }

        #endregion

        #region Matrix manipulation functions

        public int GetElement(int line, int column)
        {
            return this.matrix[line, column];
        }

        public void SetElement(int line, int column, int value)
        {
            this.matrix[line, column] = value;
        }

        public void Transpose()
        {
            int[,] rezult = new int[Size,Size];
            for (int i = 0; i < Size; i++)
                for (int j = 0; j < Size; j++)
                    rezult[i, j] = matrix[j, i];
            matrix = rezult;
        }

        public void Sort()
        {
            int[] temparray = new int[Size];
            for (int i = 0; i < Size; i++)
            {
                Array.Copy(matrix, Size*i, temparray, 0, Size);
                temparray = temparray.OrderByDescending(c => c).ToArray();
                Array.Copy(temparray, 0, matrix, Size*i, Size);
            }
        }

        public void Print()
        {
            for (int i = 0; i < Size; i++)
            {
                for (int j = 0; j < Size; j++)
                    Console.Write("{0,6d}  ", matrix[i, j]);
                Console.WriteLine();
            }
        }

        #endregion
    }
}

File: lab5/lab5/Matrix.cpp
#include "Data.h"
#include <algorithm>
#include <iostream>

Matrix::Matrix(int sizes){
	size = sizes;
	matrix = new int*[size];
	for (int i = 0; i < size; ++i)
		matrix[i] = new int[size];
}

void Matrix::generate(int filledDigit){
	for (int i = 0; i < size; i++)
		for (int j = 0; j < size; j++)
			matrix[i][j] = filledDigit;
}

Matrix* Matrix::operator*(Matrix &matr){
	Matrix *rez = new Matrix(size);
	int tempValue = 0;
	for (int i = 0; i < size; i++)
	for (int j = 0; j < size; j++) {
		tempValue = 0;
		for (int k = 0; k < size; k++)
			tempValue += matrix[j][k] * matr.getElement(k, j);
		rez->setElement(i, j, tempValue);
	}
	return rez;
}

void Matrix::print(){
	for (int i = 0; i < size; i++){
		for (int j = 0; j < size; j++)
			std::printf("%6d  ", matrix[i][j]);
		std::printf("\n");
	}
}

Matrix* Matrix::operator+(Matrix &matr){
	Matrix *rezult = new Matrix(size);
	for (int i = 0; i < size; i++)
	for (int j = 0; j < size; j++)
		rezult->setElement(i, j, matrix[i][j] + matr.getElement(i, j));
	return rezult;
}

Matrix* Matrix::sort(){
	Matrix *rezult = new Matrix(size);
	std::copy(matrix, matrix + size*size, (int*)rezult + 1);
	for (int i = 0; i < size; i++)
		qsort((int*)rezult+1+i*size, size, sizeof(int), [](const void *a, const void *b){return *((int*)b) - *((int*)a); });
	return rezult;
}

Matrix* Matrix::transpose(){
	Matrix *rezult = new Matrix(size);
	for (int i = 0; i < size; i++)
	for (int j = 0; j < size; j++)
		rezult->setElement(i, j, matrix[j][i]);
	return rezult;
}

int Matrix::getSize(){
	return size;
}

int Matrix::getElement(int line, int column){
	return matrix[line][column];
}
void Matrix::setElement(int line, int column, int value){
	matrix[line][column] = value;
}

Matrix::~Matrix()
{
	for (int i = 0; i < size; i++)
		delete[] matrix[i];
	delete[] matrix;
}

Vector* operator*(Matrix &matrix, Vector &vector){
	Vector *rez = new Vector(vector.getSize());
	int tempValue = 0;
	for (int i = 0; i < vector.getSize(); i++){
		tempValue = 0;
		for (int j = 0; j < vector.getSize(); j++)
			tempValue += vector.getElement(i) * matrix.getElement(j, i);
		rez->setElement(i, tempValue);
	}
	return rez;
}

Vector* operator*(Vector &vector, Matrix &matrix){
	Vector *rez = new Vector(vector.getSize());
	int tempValue = 0;
	for (int i = 0; i < vector.getSize(); i++){
		tempValue = 0;
		for (int j = 0; j < vector.getSize(); j++)
			tempValue += vector.getElement(i) * matrix.getElement(j, i);
		rez->setElement(i, tempValue);
	}
	return rez;
}